File: android/src/main/java/com/taqlyn/HybridTaqlynSdk.kt
package com.taqlyn

import android.content.Context
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise
import com.margelo.nitro.taqlyn.DeferredLinkPayload
import com.margelo.nitro.taqlyn.HybridTaqlynSdkSpec
import com.taqlyn.bridge.TaqlynSdkCoreBridge
import com.taqlyn.bridge.WireDeferredLink
import java.util.HashMap

/**
 * Nitro Hybrid → [TaqlynSdkCoreBridge] → Android SdkCore.
 * Matching / Install Referrer stay in SdkCore.
 */
@DoNotStrip
@Keep
class HybridTaqlynSdk : HybridTaqlynSdkSpec() {
  override fun configure(
    clientId: String,
    publicKeyId: String,
    apiBaseUrl: String,
    linkProcessingMode: String,
    env: String,
  ) {
    TaqlynSdkCoreBridge.configure(
      clientId = clientId,
      publicKeyId = publicKeyId,
      apiBaseUrl = apiBaseUrl,
      linkProcessingMode = linkProcessingMode,
      env = env,
      context = resolveAppContext(),
    )
  }

  override fun resolveDeferred(): Promise<DeferredLinkPayload?> =
    Promise.async {
      TaqlynSdkCoreBridge.resolveDeferred()?.toPayload()
    }

  override fun addLinkListener(listener: (link: DeferredLinkPayload) -> Unit) {
    TaqlynSdkCoreBridge.observeLinks { wire ->
      listener(wire.toPayload())
    }
  }

  override fun removeLinkListener() {
    TaqlynSdkCoreBridge.stopObserving()
  }

  override fun consume(linkId: String) {
    TaqlynSdkCoreBridge.consume(linkId)
  }

  override fun setReadyForNavigation(ready: Boolean) {
    TaqlynSdkCoreBridge.setReadyForNavigation(ready)
  }

  private fun resolveAppContext(): Context {
    TaqlynSdkPackage.appContext?.let { return it }
    val at = Class.forName("android.app.ActivityThread")
    val app = at.getMethod("currentApplication").invoke(null) as? Context
    return requireNotNull(app?.applicationContext) {
      "Android Context required for SdkCore.configure — ensure TaqlynSdkPackage is linked"
    }
  }

  private fun WireDeferredLink.toPayload(): DeferredLinkPayload =
    DeferredLinkPayload(
      url = url,
      path = path,
      params = HashMap(params),
      linkId = linkId,
      matchType = matchType,
      isDeferred = isDeferred,
      campaignJson = campaignJson,
    )
}

File: ios/HybridTaqlynSdk.swift
import Foundation
import NitroModules
import TaqlynSDK

/**
 Nitro Hybrid → `TaqlynSdkCoreBridge` → iOS `SdkCore`.
 Matching / pasteboard stay in TaqlynSDK.SdkCore.
 */
class HybridTaqlynSdk: HybridTaqlynSdkSpec {
  func configure(
    clientId: String,
    publicKeyId: String,
    apiBaseUrl: String,
    linkProcessingMode: String,
    env: String
  ) throws {
    _ = apiBaseUrl
    TaqlynSdkCoreBridge.configure(
      clientId: clientId,
      publicKeyId: publicKeyId,
      linkProcessingMode: linkProcessingMode,
      env: env
    )
  }

  func resolveDeferred() throws -> Promise<DeferredLinkPayload?> {
    return Promise.async {
      guard let wire = await TaqlynSdkCoreBridge.resolveDeferred() else { return nil }
      return DeferredLinkPayload(
        url: wire.url,
        path: wire.path,
        params: wire.params,
        linkId: wire.linkId,
        matchType: wire.matchType,
        isDeferred: wire.isDeferred,
        campaignJson: wire.campaignJson
      )
    }
  }

  func addLinkListener(listener: @escaping (DeferredLinkPayload) -> Void) throws {
    TaqlynSdkCoreBridge.observeLinks { wire in
      listener(
        DeferredLinkPayload(
          url: wire.url,
          path: wire.path,
          params: wire.params,
          linkId: wire.linkId,
          matchType: wire.matchType,
          isDeferred: wire.isDeferred,
          campaignJson: wire.campaignJson
        )
      )
    }
  }

  func removeLinkListener() throws {
    TaqlynSdkCoreBridge.stopObserving()
  }

  func consume(linkId: String) throws {
    TaqlynSdkCoreBridge.consume(linkId)
  }

  func setReadyForNavigation(ready: Bool) throws {
    TaqlynSdkCoreBridge.setReadyForNavigation(ready)
  }
}
